proto/proxy/destination/destination.proto

```diff
@@ -10,12 +10,40 @@ import "common/common.proto";
 // weighted set of addresses and address metadata.  Can be implemented with DNS
 // or lookups against other service discovery backends.
 //
+// If the service does not exist then the controller must send
+// `no_endpoints{exists: false}` ASAP when a client subscribes or when the
+// service stops existing. If the service exists and has endpoints available
+// then the controller must send `add` that lists all (or at least a large
+// number) of the endpoints for the service. If and only if the service exists
+// but does not have any endpoints available then the controller SHOULD send
+// `no_endpoints{exists: true}` when a client subscribes. In other words, the
+// `no_endpoints` message must only be sent when there are *no*endpoints for
+// the service.
+//
 // The controller is expected to send an Update every time there is a
 // change in service discovery.  The controller is also expected to send an
 // update at least once every ADDRESS_UPDATE_INTERVAL to indicate that the
 // controller is still healthy.  If no service discovery updates have taken
 // place, the controller can simply send an empty `add`.  The controller may
 // determine the value of ADDRESS_UPDATE_INTERVAL.
+//
+// The client MUST be prepared to receive messages in any order and the client
```

Comment:
> Is it even possible for the client to be able to handle out-of-order messages? [Add(1.1.1.1); Remove(1.1.1.1)] has different semantics from [Remove(1.1.1.1); Add(1.1.1.1)].

You are right; in the former case the client must calculate the set as empty; in the latter case the client must calculate the set as {1.1.1.1}. That's what the client does now. In practice the server probably won't send sequences where something is `Remove`d that isn't in the set. I've chosen to require the client to handle that case just because it's straightforward to do and there's no real good alternative thing for the client to do when it gets a nonsensical order.